internal/sql/pg/tx.go

```diff
@@ -114,3 +114,65 @@ func (tx *readTx) Rollback(ctx context.Context) error {
 
 	return tx.nestedTx.Rollback(ctx)
 }
+
+// delayedReadTx is a tx that handles a read-only transaction.
+// It is delayed, meaning that the tx will only be actually started
+// when the first query is executed. This is useful for when a calling
+// module is expected to control the lifetime of a read transaction, but
+// the implementation might not need to use the transaction.
+type delayedReadTx struct {
+	db *DB
+
+	tx *readTx
+}
+
+func (d *delayedReadTx) ensureTx() error {
+	if d.tx == nil {
+		tx, err := d.db.BeginReadTx(context.Background())
```

Comment:
Sorry, last thing.  Should we pass the context into `ensureTx`?  I was noting that `BeginDelayedReadTx` does not take a context for clear reasons (it doesn't do anything to cancel), but it seems that the `begin transaction ....` bits should be cancelable as well.

Reply:
Ah yeah. Not sure why I did that, we definitely want to pass the context there.